charmcraft/charm_builder.py

```diff
@@ -0,0 +1,357 @@
+                            raise
+                        shutil.copy2(str(abs_path), str(dest_path))
+                else:
+                    logger.debug("Ignoring file because of type: %r", str(rel_path))
+
+        # the linked entrypoint is calculated here because it's when it's really in the build dir
+        linked_entrypoint = self.buildpath / self.entrypoint.relative_to(self.charmdir)
+
+        return linked_entrypoint
+
+    def handle_dispatcher(self, linked_entrypoint):
+        """Handle modern and classic dispatch mechanisms."""
+        # dispatch mechanism, create one if wasn't provided by the project
+        dispatch_path = self.buildpath / DISPATCH_FILENAME
+        if not dispatch_path.exists():
+            logger.debug("Creating the dispatch mechanism")
+            dispatch_content = DISPATCH_CONTENT.format(
+                entrypoint_relative_path=linked_entrypoint.relative_to(self.buildpath)
+            )
+            with dispatch_path.open("wt", encoding="utf8") as fh:
+                fh.write(dispatch_content)
+                make_executable(fh)
+
+        # bunch of symlinks, to support old juju: verify that any of the already included hooks
+        # in the directory is not linking directly to the entrypoint, and also check all the
+        # mandatory ones are present
+        dest_hookpath = self.buildpath / HOOKS_DIR
+        if not dest_hookpath.exists():
+            dest_hookpath.mkdir()
+
+        # get those built hooks that we need to replace because they are pointing to the
+        # entrypoint directly and we need to fix the environment in the middle
+        current_hooks_to_replace = []
+        for node in dest_hookpath.iterdir():
+            if node.resolve() == linked_entrypoint:
+                current_hooks_to_replace.append(node)
+                node.unlink()
+                logger.debug(
+                    "Replacing existing hook %r as it's a symlink to the entrypoint",
+                    node.name,
+                )
+
+        # include the mandatory ones and those we need to replace
+        hooknames = MANDATORY_HOOK_NAMES | {x.name for x in current_hooks_to_replace}
+        for hookname in hooknames:
+            logger.debug("Creating the %r hook script pointing to dispatch", hookname)
+            dest_hook = dest_hookpath / hookname
+            if not dest_hook.exists():
+                relative_link = relativise(dest_hook, dispatch_path)
+                dest_hook.symlink_to(relative_link)
+
+    def handle_dependencies(self):
+        """Handle from-directory and virtualenv dependencies."""
+        logger.debug("Installing dependencies")
+
+        if not self.python_packages and not self.requirements:
+            return
+
+        # create venv so packages can be cached. note: additional packages
+        # pip and setuptools will also be installed!
+        staging_venv_dir = self.charmdir / STAGING_VENV_DIRNAME
```

Comment:
Note that here you're adding stuff to the project's directory, not inside the building one. This is something we never did so far.

Reply:
This is a temporary condition that exists only while the builder tool is invoked directly by the build command without the parts lifecycle. In this case, origin is charmdir and destination is buildpath. When invoked from the parts lifecycle, origin will be `<work_dir>/parts/charm/build` and destination will be `<work_dir>/parts/charm/install`, and the directory will not be created in charmdir.